### PR TITLE
[In-memory fan-out] EventsBuffer implementation

### PR DESCRIPTION
**Pull request by tudor-da**
Implements an ordered-by-offset ring buffer wrapper necessary for the in-memory fan-out mechanism for Ledger API serving. The implementation is generic in order to allow particularization for both TransactionsBuffer and CommandCompletionsBuffer.

CHANGELOG_BEGIN
CHANGELOG_END

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger/participant-integration-api/src/main/scala/platform/store/cache/EventsBuffer.scala [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+    *              Must be higher than the last appended entry's offset, with the exception
+    *              of the range end marker, which can have an offset equal to the last appended element.
+    * @param entry The buffer entry.
+    */
+  def push(offset: O, entry: E): Unit =
+    Timed.value(
+      pushTimer,
+      synchronized {
+        if (bufferRef.size == maxBufferSize)
+          bufferRef = bufferRef.drop(1)
```
Comment (kamil-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/cache/EventsBuffer.scala:60]: This is probably a minor thing but in case when we have a full buffer and a range-end marker is pushed, a single buffer element will be dropped but no new element will be appended. Is that intentional?

File: ledger/participant-integration-api/src/main/scala/platform/store/cache/EventsBuffer.scala [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+    * @return The series of events as an ordered vector satisfying the input bounds.
+    */
+  def slice(startExclusive: O, endInclusive: O): Vector[(O, E)] =
+    Timed.value(
+      sliceTimer, {
+        val buffer = bufferRef
+        if (buffer.isEmpty) {
+          Vector.empty
+        } else if (bufferRangeEnd.exists(_ < endInclusive)) {
+          throw RequestOffBufferBounds(buffer.last._1, endInclusive)
```
Comment (kamil-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/cache/EventsBuffer.scala:96]: Shouldn't the condition on `endInclusive` not being greater than `bufferRangeEnd` be before returning an empty buffer to have consistent behavior?
E.g. after pruning the whole buffer and asking for an out-of-range end offset we won't get an exception but an empty buffer.

File: ledger/participant-integration-api/src/test/suite/scala/platform/store/cache/EventsBufferSpec.scala [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+        buffer.slice(0, LastOffset) shouldBe Vector(8 -> 16, 13 -> 26)
+      }
+    }
+
+    "element inserted" should {
+      "prune inclusive" in withBuffer() { buffer =>
+        buffer.prune(6)
+        buffer.slice(0, LastOffset) shouldBe Vector(8 -> 16, 13 -> 26)
+      }
+    }
```
[the remark this answers is not in the document]
Comment (tudor-da) [reply ledger/participant-integration-api/src/test/suite/scala/platform/store/cache/EventsBufferSpec.scala:119]: Yes, it is basically asserting the end inclusive behavior: whether it lands on a vacant offset or not

File: ledger/participant-integration-api/src/main/scala/platform/store/cache/EventsBuffer.scala (hunk 1 above)
Comment (tudor-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/cache/EventsBuffer.scala:60]: Good catch, thanks! Fortunately, this is harmless. I'll commit a fix

File: ledger/participant-integration-api/src/test/suite/scala/platform/store/cache/EventsBufferSpec.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (kamil-da) [reply ledger/participant-integration-api/src/test/suite/scala/platform/store/cache/EventsBufferSpec.scala:119]: How about changing `element inserted` to `element not present`? I find it a bit counter-intuitive because after reading the test case header I went for looking where element `6` was inserted :)

File: ledger/participant-integration-api/src/main/scala/platform/store/cache/EventsBuffer.scala (hunk 2 above)
Comment (tudor-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/cache/EventsBuffer.scala:96]: You are right. For the sake of consistency, they should be inverted.

**Review by kamil-da — APPROVED** [review]
👏

File: ledger/participant-integration-api/src/test/suite/scala/platform/store/cache/EventsBufferSpec.scala [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+            val expected = ((idx + 1) to (100 + idx)).map(idx => idx -> idx)
+            buffer.push(1000 + idx, 1000 + idx)
+            buffer.slice(idx, 100 + idx) should contain theSameElementsInOrderAs expected
+        }
+      }
+    }
+  }
+
+  "prune" when {
+    "element found" should {
```
Comment (meiersi-da) [inline ledger/participant-integration-api/src/test/suite/scala/platform/store/cache/EventsBufferSpec.scala:107]: I could not find a test that would test that the binary-search implementation works correctly for arrays of size 0, 1, 2, which I'd consider boundary conditions for it.

Consider either testing that indirectly via `prune` here, or also copying the unit tests from the source you've copied the binary-search implementation.

**Review by meiersi-da — CHANGES_REQUESTED** [review]
@tudor-da : thanks for the ping this morning. There might be a data race lurking in the code. See the comments inline.

File: ledger/participant-integration-api/src/test/suite/scala/platform/store/cache/EventsBufferSpec.scala [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+
+  "getEvents" when {
+    "called with valid range" should {
+      val bufferElements = increasinglySparse(10).map(offset => offset -> offset * 2)
+
+      "return the correct series" in withBuffer(10, bufferElements) { buffer =>
+        val rangesGen = for {
+          startExclusive <- Gen.choose(0, bufferElements.last._1)
+          endInclusive <- Gen.choose(startExclusive, bufferElements.last._1)
+        } yield (startExclusive, endInclusive)
```
Comment (mziolekda) [inline ledger/participant-integration-api/src/test/suite/scala/platform/store/cache/EventsBufferSpec.scala:73]: I am not sure how exhaustive those generators are, I would be very interested in a pinpointed test that explicitly hits the buffer at its extremes i.e. returns the full buffer

File: ledger/participant-integration-api/src/main/scala/platform/store/cache/EventsBuffer.scala [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+    *
+    * @param startExclusive The start exclusive bound of the requested range.
+    * @param endInclusive The end inclusive bound of the requested range.
+    * @return The series of events as an ordered vector satisfying the input bounds.
+    */
+  def slice(startExclusive: O, endInclusive: O): Vector[(O, E)] =
+    Timed.value(
+      sliceTimer, {
+        val buffer = bufferRef
+        if (bufferRangeEnd.exists(_ < endInclusive)) {
```
[the remark this answers is not in the document]
Comment (tudor-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/cache/EventsBuffer.scala:92]: Yes, you are right. This method was initially synchronized, but decided that it is not needed and forgot to addthe volatile to the field. Thanks!

File: ledger/participant-integration-api/src/main/scala/platform/store/cache/EventsBuffer.scala [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+          // Ensure vector grows with strictly monotonic offsets.
+          // Only specially-designated range end markers are allowed
+          // to have offsets equal to the buffer range end.
+          if (lastOffset > offset || (lastOffset == offset && !isRangeEndMarker(entry))) {
+            throw UnorderedException(lastOffset, offset)
+          }
+        }
+
+        // Update the buffer range end
+        bufferRangeEnd = Some(offset)
```
[the remark this answers is not in the document]
Comment (tudor-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/cache/EventsBuffer.scala:68]: Yep are right. Between two changes, I dropped all memory syncs on `bufferRangeEnd`. Will update.. 
The read-write conflict can not happen in the scope of this design (the write is always ahead of the read by design), but I agree that it should be defensively implemented in here as well.

File: ledger/participant-integration-api/src/main/scala/platform/store/cache/EventsBuffer.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (meiersi-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/cache/EventsBuffer.scala:92]: Note that a volatile field on its own is not enough to avoid violating the invariant that `buffer` contains all elements up to `bufferRangeEnd`. See https://github.com/digital-asset/daml/pull/9775#discussion_r640421507

[the remark this answers is not in the document]
Comment (tudor-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/cache/EventsBuffer.scala:92]: Thanks for spotting the issues. As mentioned above, the invariant was not necessary to hold in this design since the buffer is only updated _ahead_ of the read requests. But you are right, @meiersi-da that we can easily make this invariant hold without external assumptions.

File: ledger/participant-integration-api/src/test/suite/scala/platform/store/cache/EventsBufferSpec.scala (hunk 5 above)
Comment (tudor-da) [reply ledger/participant-integration-api/src/test/suite/scala/platform/store/cache/EventsBufferSpec.scala:73]: Fine with that.

File: ledger/participant-integration-api/src/test/suite/scala/platform/store/cache/EventsBufferSpec.scala (hunk 4 above)
Comment (tudor-da) [reply ledger/participant-integration-api/src/test/suite/scala/platform/store/cache/EventsBufferSpec.scala:126]: Good idea, why not: Wrote a small suite for that binary search method

File: ledger/participant-integration-api/src/main/scala/platform/store/cache/EventsBuffer.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (tudor-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/cache/EventsBuffer.scala]: Done. Created a `BufferStateRef` wrapper case class for the two values holding the state, so it can be easily addressed and updated atomically.

File: ledger/participant-integration-api/src/test/suite/scala/platform/store/cache/EventsBufferSpec.scala (hunk 5 above)
Comment (tudor-da) [reply ledger/participant-integration-api/src/test/suite/scala/platform/store/cache/EventsBufferSpec.scala]: Done.

File: ledger/metrics/src/main/scala/com/daml/metrics/Metrics.scala [hunk 8]
```diff
@@ -636,6 +636,14 @@ final class Metrics(val registry: MetricRegistry) {
         val deduplicateCommand: Timer = registry.timer(Prefix :+ "deduplicate_command")
         val stopDeduplicateCommand: Timer = registry.timer(Prefix :+ "stop_deduplicating_command")
         val prune: Timer = registry.timer(Prefix :+ "prune")
+
+        object buffer {
```
[the remark this answers is not in the document]
Comment (tudor-da) [reply ledger/metrics/src/main/scala/com/daml/metrics/Metrics.scala]: Done

**Review by meiersi-da — APPROVED** [review]
Thanks for fixing all the comments @tudor-da ! Much appreciated!
